Remove empty test suite `tests/run-make-fulldeps`

Zalathar opened the pull request:
After #109770, there were only a handful of tests left in the run-make-fulldeps suite.

As of #126111, there are no longer *any* run-make-fulldeps tests, so now we can:

- Remove the directory
- Remove related bootstrap/compiletest code
- Remove various other references in CI scripts and documentation.

By removing this suite, we also no longer need to worry about discrepancies between it and ui-fulldeps, and we don't have to worry about porting tests from Makefile to [rmake](https://github.com/rust-lang/rust/issues/121876) (or whether rmake even works with fulldeps).

Zalathar committed:
Remove empty test suite `tests/run-make-fulldeps`

Zalathar committed:
Identify run-make tests by mode name instead of suite name

Zalathar committed:
Remove useless feature gate test for `#[feature(extern_prelude)]`

This test never actually checked anything useful, so presumably it only existed
to silence the tidy check for feature gate tests, with the real checks being
performed elsewhere (in tests that have since been deleted).

Zalathar commented on src/ci/docker/host-x86_64/x86_64-gnu-debug/Dockerfile:
```diff
@@ -44,6 +44,4 @@ ENV RUST_CONFIGURE_ARGS \
       --set target.x86_64-unknown-linux-gnu.cc=clang \
       --set target.x86_64-unknown-linux-gnu.cxx=clang++
 
-ENV SCRIPT \
-  python3 ../x.py --stage 2 build && \
-  python3 ../x.py --stage 2 test tests/run-make-fulldeps --test-args clang
```
This one is a bit of a head-scratcher for me, because with the removal of run-make-fulldeps it's unclear what this job is actually supposed to be testing.

Zalathar replied:
The line that tests run-make-fulldeps was originally added in #57514 (specifically https://github.com/rust-lang/rust/pull/57514/commits/48cb04fbca1d398c170b713b3fe7e681a8e39a4f).

Kobzol replied:
Hmm, this image wasn't actually running any tests for some time, because there were no tests marked with `needs-matching-clang` left in the `run-make-fulldeps` directory.

Kobzol replied:
I think that we should just change this invocation to `tests/run-make`, for now. In the longer term, I'm planning to document and inspect all our CI jobs to make sure that we don't miss any tests, and that we don't do useless work, because apparently some of them are very old :)

Zalathar replied:
Looking at the history of this job, I think it was originally intended to just build the compiler with `--enable-debug`, to check that the build succeeds.

Then the clang-specific stuff was grafted on later.

Zalathar replied:
But since the clang-specific stuff has been moved over into run-make, replacing this with run-make still makes sense, probably.

kadiwa4 commented on src/doc/rustc/src/platform-support/nto-qnx.md:
```diff
@@ -161,7 +161,6 @@ export exclude_tests='
     --exclude tests/ui-fulldeps
     --exclude rustc
     --exclude rustdoc
-    --exclude tests/run-make-fulldeps'
```
you removed the closing single quote

Zalathar replied:
Good catch, thanks.

onur-ozkan reviewed (approved):
Looks good to me, thanks!

r=me once https://github.com/rust-lang/rust/pull/126155#discussion_r1632057966 is addressed.

Kobzol commented on src/ci/github-actions/jobs.yml:
```diff
@@ -37,7 +37,7 @@ runners:
 envs:
   env-x86_64-apple-tests: &env-x86_64-apple-tests
-    SCRIPT: ./x.py --stage 2 test --skip tests/ui --skip tests/rustdoc --skip tests/run-make-fulldeps
+    SCRIPT: ./x.py --stage 2 test --skip tests/ui --skip tests/rustdoc
```
This removes test coverage from Apple, in theory. I wonder if we should instead change it to `--skip tests/run-make` and add `tests/run-make` to `x86_64-apple-2`. Although some of the tests were actually moved to `ui-fulldeps` :thinking:

Zalathar replied:
From what I can tell, the Apple test suites are split between two jobs: one that skips (ui, rustdoc, run-make-fulldeps), and one that only runs those jobs.

So I don't think this actually affects overall test coverage on Apple?

Kobzol replied:
Yes, this was preexisting. When the tests moved from -fulldeps to just run-make, we stopped testing them on Apple.

Kobzol commented on src/bootstrap/src/core/build_steps/test.rs:
```diff
@@ -1973,9 +1967,7 @@ NOTE: if you're sure you want to do this, please open an issue as to why. In the
-            if !builder.config.dry_run()
-                && (matches!(suite, "run-make" | "run-make-fulldeps") || mode == "coverage-run")
-            {
+            if !builder.config.dry_run() && matches!(mode, "run-make" | "coverage-run") {
```
This now checks the `mode` for `run-make`, instead of the `suite`. Is it the same?

jieyouxu replied:
I think this is okay, because the test mode is "run-make" but the test suite name is also "run-make".

Zalathar committed:
Make job `x86_64-gnu-debug` run a subset of run-make tests

It looks like this job was intending to run all of the `needs-matching-clang`
tests (since they don't run without `RUSTBUILD_FORCE_CLANG_BASED_TESTS`), but
over time developed two problems:

- The tests it cares about were moved from run-make-fulldeps to run-make.
- Some of the relevant tests don't actually have "clang" in their name.

Switching to run-make solves the first problem, but we still don't run the
tests without "clang" in their name, because some of them are currently broken.

Zalathar commented:
I made the suggested changes to `x86_64-gnu-debug` [(diff)](https://github.com/rust-lang/rust/compare/f5139e77c30bea3bf430b034162cb0c390018bdd..8a3067f69bcd0e4880fed4410285969d002971ca), and also added a brief comment with my current (imperfect) understanding of what this job is trying to do.

Note that I have removed the `--test-args clang` part, because it was failing to run some tests that use the `//@ needs-matching-clang` header but don't have `clang` in their name. Running the whole suite is overkill, but it seems preferable to accidentally missing some tests that otherwise wouldn't run anywhere.

Zalathar commented:
I've also set up a run of the debug job over in #114917, because it looks like some of the affected tests haven't actually been running for a while, so I want to see whether any of them fail.

Zalathar commented:
[Looks like my test run found some tests that are broken:](https://github.com/rust-lang/rust/pull/114917#issuecomment-2156313115)

```text
      [run-make] tests/run-make/cross-lang-lto-pgo-smoketest
      [run-make] tests/run-make/cross-lang-lto-riscv-abi
      [run-make] tests/run-make/issue-84395-lto-embed-bitcode
```

In light of this, I'll update the Dockerfile to continue passing `--test-args clang`, and we can follow up these other tests in a separate issue.

onur-ozkan commented on src/ci/github-actions/jobs.yml:
```diff
@@ -294,7 +294,7 @@ auto:
   - image: x86_64-apple-2
     env:
-      SCRIPT: ./x.py --stage 2 test tests/ui tests/rustdoc tests/run-make-fulldeps
+      SCRIPT: ./x.py --stage 2 test tests/ui tests/rustdoc
```
Shouldn't we run `run-make` tests here?

cc @Kobzol

Kobzol replied:
We should be running them [here](https://github.com/rust-lang/rust/pull/126155/files#diff-ef98dea3810f7940c389a6d469a0fad6c21da049c0ad16e9b0f27768c87dce2cR40), since that runs all tests except for ui and rustdoc.

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit 5223bf4474b02bd6925c5b516a8a0efd5ec4ec20 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 5223bf4474b02bd6925c5b516a8a0efd5ec4ec20 -->
<!-- homu: {"type":"Approved","sha":"5223bf4474b02bd6925c5b516a8a0efd5ec4ec20","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
